internal/fetch/fetch.go

```diff
@@ -158,22 +159,32 @@ func DownloadTarball(target, url, expectedSha256 string) error {
 
 // downloadTarball downloads a tarball from the given source URL to the target
 // path. It retries up to 3 times with exponential backoff on failure.
-func downloadTarball(target, source string) error {
+func downloadTarball(ctx context.Context, target, source string) error {
 	var err error
 	backoff := 10 * time.Second
 	for i := range 3 {
-		if i != 0 {
-			time.Sleep(backoff)
-			backoff = 2 * backoff
+		if i > 0 {
+			select {
+			case <-time.After(backoff):
+				backoff *= 2
+			case <-ctx.Done():
+				return ctx.Err()
+			}
 		}
-		if err = downloadAttempt(target, source); err == nil {
+
+		err = downloadAttempt(ctx, target, source)
```

Comment:
nit: In Go, it's more idiomatic to check for errors (err != nil) rather than success. I would do:


if err := downloadAttempt(ctx, target, source); err != nil {
    if errors.Is(err, context.Canceled) {
        return err
    }
    continue
}
return nil